Onboarding Customisation Experiment

nalcalag opened the pull request:
Task/Issue URL: https://app.asana.com/0/72649045549333/1203826977705495/f

### Description
Implement Onboarding Customisation experiment where experimental group will see an extra onboarding dialog with DDG feature options.

### Steps to test this PR

_Change all variants to 0 except for the `mj` variant_

_New multiselect dialog_
- Uninstall previous version of the app or clear storage
- Install from this branch
- Welcome animation will appear
- [ ] Check new dialog with DDG feature options appear

_Skip selecting options_
- Follow same instructions as "New multiselect dialog" test
- Tap on `Skip This Step`
- [ ] Check existing welcome cta dialog appears
- [ ] Check pixel `m_onboarding_option_skip` is fired

_Continue without options selected_
- Follow same instructions as "New multiselect dialog" test
- Tap on any option to select it
- Remove the selection tapping again
- Tap on `Continue` button
- [ ] Check existing welcome cta dialog appears
- [ ] Check pixel `m_onboarding_option_skip` is fired

_Continue with options selected_
- Follow same instructions as "New multiselect dialog" test
- Tap on one or more options to select them
- Tap on `Continue` button
- [ ] Check existing welcome cta dialog appears
- [ ] Check pixel `m_onboarding_option_selected` is fired
- [ ] Check pixels for each option selected is also fired. You can find the list of pixels for each option here -> https://app.asana.com/0/0/1204137437551040/f

### UI changes


https://user-images.githubusercontent.com/20798495/228512221-df3caf98-2bda-4f0e-9976-adfc51776f9e.mp4

nalcalag commented:
Current dependencies on/for this PR:
* develop
  * **PR #3003** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3003" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3003?utm_source=stack-comment).

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePage.kt:
```diff
@@ -189,14 +250,20 @@ class WelcomePage : OnboardingPageFragment(R.layout.content_onboarding_welcome)
+    }
+
+    private fun showDaxDialogCta() {
+        binding.daxDialogCta.root.show()
```
same question about `root` here. Is this something related to having a `include` in the layout?

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt:
```diff
@@ -33,16 +42,52 @@ class WelcomePageViewModel(
+    private fun onContinueWithOptionsClicked(options: Map<DDGFeatureOnboardingOption, Boolean>): Flow<WelcomePageView.State> = flow {
+        val optionsSelected = options.filter { it.value }.map { it.key }.toList()
+        if (optionsSelected.isEmpty()) {
+            pixel.fire(AppPixelName.ONBOARDING_OPTION_SKIP)
```
I thought this was not possible, but testing I see it's possible to select and unselect again any option, leaving the primary button enabled without anything selected. I assume we want that, but sharing I would have expected the primary to be disabled again.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt:
```diff
@@ -33,16 +42,52 @@ class WelcomePageViewModel(
+        emit(WelcomePageView.State.ShowControlDaxCta)
+    }
+
+    private fun onContinueWithOptionsClicked(options: Map<DDGFeatureOnboardingOption, Boolean>): Flow<WelcomePageView.State> = flow {
```
NIT: if we ever decide to release the experiment (I don't think is the plan), I would move all the pixel logic into its own class.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt:
```diff
@@ -40,6 +40,15 @@ enum class AppPixelName(override val pixelName: String) : Pixel.PixelName {
     ONBOARDING_DAX_CTA_OK_BUTTON("m_odc_ok"),
     ONBOARDING_PRIVACY_SHIELD_BUTTON("m_onboarding_privacy_shield_button"),
 
+    ONBOARDING_OPTION_PRIVATE_SEARCH_SELECTED("m_onboarding_option_private_search_selected"),
```
did we open a privacy triage for these ones?

nalcalag replied:
yes :)

cmonfortep commented on app/src/main/res/layout/include_dax_dialog_cta.xml:
probably not part of the experiment and we can discard changes here?

nalcalag replied:
💯

cmonfortep commented on app/src/main/res/layout/include_dax_multiselect_dialog_cta.xml:
```diff
@@ -0,0 +1,146 @@
+    android:alpha="0"
+    android:clipChildren="false"
+    android:orientation="vertical"
+    android:visibility="gone">
```
NIT: I would avoid setting visibility to something we can include somewhere else.

nalcalag replied:
You are right, I will move this logic to the class 💯

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt:
```diff
@@ -23,17 +23,33 @@ import com.duckduckgo.app.global.DispatcherProvider
+    private val command = Channel<Command>(1, DROP_OLDEST)
+    internal fun commands(): Flow<Command> = command.receiveAsFlow()
+
+    fun determineScreenOrientation() {
```
include one test for this?

nalcalag replied:
I will 👌

cmonfortep reviewed (approved):
LGTM

works as expected, leaving the UI discussion to the UI code owners 😁.

(Agree with David that new components should go through ADS standard process, however not clear if this is an established component atm. I assume we have some lint checks and that's why we put some colors inside common-ui, if that's the case, maybe we can do something similar to what we do with strings...have a separated file, or declare them under a comment that indicates they are not established. Just some thoughts)

nalcalag commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt:
```diff
@@ -52,4 +68,8 @@ class OnboardingViewModel @Inject constructor(
     }
+
+    internal sealed class Command {
+        object SetScreenOrientation : Command()
```
You are right, I changed it to `ForceToPortraitForMobileDevices` so it's more clear 👍

nalcalag commented:
@malmstein following @cmonfortep suggestion I added a new resource file called `temp_colors.xml` where I put all the colors needed for the customView because I need to add them to the theme file so they can change depending on the theme. I also migrated all styles and attributes to new resources files in the `app` module so they are not in `common-ui` anymore 🎉

@cmonfortep thanks a lot for all your feedback, I changed everything you suggested and added tests for portrait checks 👍

malmstein reviewed (approved):
Thanks for doing the changes @nalcalag. Good work!
